README.md
# ndcombine

WIP implementation of ND arrays combination.


CHANGES.rst
0.1 (unreleased)
----------------

First version.


test.py
import os
import sys
import time
import timeit
from pathlib import Path

import bottleneck as bn
import numpy as np
from astropy.nddata import CCDData

from ndcombine import combine_arrays, ndcombine

data = None
mask = None
datamasked = None


def test_median():
    global data, mask, datamasked

    shape = (10, 1000, 1000)
    np.random.seed(42)
    data = np.random.normal(size=shape).astype(np.float32)
    mask = np.zeros(shape, dtype=np.uint16)
    data = data.reshape(shape[0], -1)
    mask = mask.reshape(shape[0], -1)
    datamasked = np.ma.array(data, mask=mask.astype(bool))

    outndcomb, _, _ = ndcombine(data,
                                mask,
                                combine_method='median',
                                reject_method='none')
    outnp = np.median(data, axis=0)
    outnpma = np.ma.median(datamasked, axis=0)
    outbn = bn.median(data, axis=0)

    np.testing.assert_array_equal(outndcomb, outnp)
    np.testing.assert_array_equal(outndcomb, outnpma)
    np.testing.assert_array_equal(outndcomb, outbn)

    nb = 10
    kwargs = dict(globals=globals(), number=nb, repeat=5)

    def run(label, command):
        res = timeit.repeat(command, **kwargs)
        res = np.array(res) / nb
        print(res)
        print(f'{label:20s}: {np.mean(res):.3f}s ± {np.std(res):.3f}s')

    run('np.median', 'np.median(data)')
    run('np.ma.median', 'np.ma.median(datamasked, axis=0)')
    run('bn.median', 'bn.median(data)')
    run('ndcombine 1 thread',
        "ndcombine(data, mask, combine_method='median', "
        "reject_method='none', num_threads=1)")
    run('ndcombine',
        "ndcombine(data, mask, combine_method='median', "
        "reject_method='none')")


def test_files(case):
    datadir = Path(os.path.expanduser('~/data/combiner'))
    flist = list(datadir.glob('image-*.fits'))
    ccds = [CCDData.read(f) for f in flist]

    if case == 'profile':
        import line_profiler
        profile = line_profiler.LineProfiler(combine_arrays)
        profile.runcall(combine_arrays,
                        ccds,
                        method='mean',
                        clipping_method='sigclip')
        profile.print_stats()
    else:
        n = 5

        t0 = time.time()
        for _ in range(n):
            print('.', end='', flush=True)
            combine_arrays(ccds,
                           method='mean',
                           clipping_method='sigclip',
                           num_threads=0)
        print('\nMean of 5 with max threads : {:.2f} sec.'.format(
            (time.time() - t0) / n))

        t0 = time.time()
        for _ in range(n):
            print('.', end='', flush=True)
            combine_arrays(ccds,
                           method='mean',
                           clipping_method='sigclip',
                           num_threads=1)
        print('\nMean of 5 with 1 thread : {:.2f} sec.'.format(
            (time.time() - t0) / n))


if __name__ == "__main__":
    case = sys.argv[1] if len(sys.argv) > 1 else 'default'

    if case == 'median':
        test_median()
    else:
        test_files(case)


ndcombine/tests/helpers.py
# Adapted from https://github.com/astropy/astroscrappy
# (astroscrappy/tests/fake_data.py, licensed under BSD-3)

import os

import numpy as np
from astropy.nddata import CCDData, VarianceUncertainty


# Make a simple Gaussian function for testing purposes
def gaussian(image_shape, x0, y0, brightness, fwhm):
    x = np.arange(image_shape[1])
    y = np.arange(image_shape[0])
    x2d, y2d = np.meshgrid(x, y)
    sig = fwhm / 2.35482
    normfactor = brightness / 2.0 / np.pi * sig**-2.0
    exponent = -0.5 * sig**-2.0
    exponent *= (x2d - x0)**2.0 + (y2d - y0)**2.0
    return normfactor * np.exp(exponent)


def make_fake_data(nimg,
                   outdir,
                   nsources=100,
                   shape=(2048, 2048),
                   dtype=np.float32):
    # Set a seed so that the tests are repeatable
    np.random.seed(200)

    # Add some fake sources
    sources = np.zeros(shape, dtype=np.float32)
    xx = np.random.uniform(low=0.0, high=shape[0], size=nsources)
    yy = np.random.uniform(low=0.0, high=shape[1], size=nsources)
    brightness = np.random.uniform(low=1000., high=30000., size=nsources)
    for x, y, b in zip(xx, yy, brightness):
        sources += gaussian(shape, x, y, b, 5)

    for i in range(nimg):
        # Create a simulated image to use in our tests
        imdata = np.zeros(shape, dtype=dtype)
        # Add sky and sky noise
        imdata += 200

        imdata += sources

        # Add the poisson noise
        imdata = np.float32(np.random.poisson(imdata))

        # Add readnoise
        imdata += np.random.normal(0.0, 10.0, size=shape)

        # Add 100 fake cosmic rays
        cr_x = np.random.randint(low=5, high=shape[0] - 5, size=100)
        cr_y = np.random.randint(low=5, high=shape[1] - 5, size=100)
        cr_brightnesses = np.random.uniform(low=1000.0, high=30000.0, size=100)
        imdata[cr_y, cr_x] += cr_brightnesses
        imdata = imdata.astype('f4')

        # Make a mask where the detected cosmic rays should be
        # crmask = np.zeros(shape, dtype=np.bool)
        # crmask[cr_y, cr_x] = True

        ccd = CCDData(imdata,
                      uncertainty=VarianceUncertainty(imdata / 10),
                      unit="electron")
        ccd.write(os.path.join(outdir, f'image-{i+1:02d}.fits'),
                  overwrite=True)
        print('.', end='')


ndcombine/tests/test_ndcombine.py
import numpy as np
from astropy.stats import sigma_clip as sigma_clip_ast
from ndcombine import combine_arrays, sigma_clip
from numpy.testing import assert_array_equal, assert_array_almost_equal


def test_sigclip():
    data = np.array([1., 2, 3, 2, 3, 2, 1, 4, 2, 100], dtype=np.float32)
    mask1 = sigma_clip_ast(data).mask.astype(int)
    mask2 = sigma_clip(data, lsigma=3, hsigma=3, max_iters=10)
    assert_array_equal(mask1, mask2)

    mask = np.zeros_like(data, dtype=np.uint16)
    mask[7] = 1
    mask1 = sigma_clip_ast(np.ma.array(data, mask=mask)).mask.astype(int)
    mask2 = sigma_clip(data, mask=mask, lsigma=3, hsigma=3, max_iters=10)
    assert_array_equal(mask1, mask2)


def test_simple():
    data = np.array([[1., 2, 3, 2, 3, 2, 1, 4, 2, 100]],
                    dtype=np.float32).T
    out = combine_arrays(data, method='mean', clipping_method='sigclip')
    assert_array_equal(out.meta['REJMASK'].ravel(),
                       [0, 0, 0, 0, 0, 0, 0, 0, 0, 1])
    assert_array_almost_equal(out.data.ravel(), [2.2222223])


benchmarks/median.py
import timeit
import bottleneck as bn
import numpy as np
from ndcombine import ndcombine

data = None
mask = None
datamasked = None


def test_median():
    global data, mask, datamasked

    shape = (10, 1000, 1000)
    np.random.seed(42)
    data = np.random.normal(size=shape).astype(np.float32)
    mask = np.zeros(shape, dtype=np.uint16)
    data = data.reshape(shape[0], -1)
    mask = mask.reshape(shape[0], -1)
    datamasked = np.ma.array(data, mask=mask.astype(bool))

    outndcomb, _, _ = ndcombine(data,
                                mask,
                                combine_method='median',
                                reject_method='none')
    outnp = np.median(data, axis=0)
    outnpma = np.ma.median(datamasked, axis=0)
    outbn = bn.median(data, axis=0)

    np.testing.assert_array_equal(outndcomb, outnp)
    np.testing.assert_array_equal(outndcomb, outnpma)
    np.testing.assert_array_equal(outndcomb, outbn)

    nb = 10
    kwargs = dict(globals=globals(), number=nb, repeat=5)

    def run(label, command):
        res = timeit.repeat(command, **kwargs)
        res = np.array(res) / nb
        print(f'{label:20s}: {np.mean(res):.3f}s ± {np.std(res):.3f}s')

    run('np.median', 'np.median(data)')
    run('np.ma.median', 'np.ma.median(datamasked, axis=0)')
    run('bn.median', 'bn.median(data)')
    run(
        'ndcombine 1 thread', "ndcombine(data, mask, combine_method='median', "
        "reject_method='none', num_threads=1)")
    run(
        'ndcombine', "ndcombine(data, mask, combine_method='median', "
        "reject_method='none')")


if __name__ == "__main__":
    test_median()


benchmarks/combine.py
import argparse
import os
import time
import warnings
from collections import defaultdict
from functools import wraps
from pathlib import Path

import ccdproc
import imcombinepy
import numpy as np
from astrodata import NDAstroData
from astropy.nddata import CCDData
from astropy.table import Table

with warnings.catch_warnings():
    warnings.simplefilter('ignore')
    from gempy.library.nddops import NDStacker

from ndcombine import combine_arrays

benchmarks = {
    'ndcombine': {
        'mean': {'method': 'mean'},
        'mean+sigclip': {'method': 'mean', 'clipping_method': 'sigclip'},
        'median': {'method': 'median'},
    },
    'ccdproc': {
        'mean': {'method': 'average', 'sigma_clip': False},
        'mean+sigclip': {'method': 'average', 'sigma_clip': True},
        'median': {'method': 'median', 'sigma_clip': False},
    },
    'dragons': {
        'mean': {'combine': 'mean'},
        'mean+sigclip': {'combine': 'mean', 'reject': 'sigclip'},
        'median': {'combine': 'median'},
    },
    'imcombinepy': {
        'mean': {'combine': 'mean'},
        'mean+sigclip': {'combine': 'mean', 'reject': 'sigclip'},
        'median': {'combine': 'median'},
    },
}


def time_execution(f):
    """Decorator which returns the execution time of a function."""

    @wraps(f)
    def timed(*args, **kw):
        t0 = time.time()
        n = 5
        for _ in range(n):
            f(*args, **kw)
            print('.', end='', flush=True)

        measured = (time.time() - t0) / n
        print(f'\nMean of 5 : {measured:.2f} sec.')
        return measured

    return timed


class Compare:

    def __init__(self, limit=None, datadir='~/data/combiner'):
        datadir = Path(os.path.expanduser(datadir))

        if not datadir.exists():
            print('Creating test data')
            # datadir.mkdir(parents=True)
            from ndcombine.tests.helpers import make_fake_data
            make_fake_data(20, datadir, nsources=500)

        flist = list(datadir.glob('image-*.fits'))
        self.ccds = [CCDData.read(f, memmap=False) for f in flist[:limit]]
        self.ndds = [
            NDAstroData(ccd.data, uncertainty=ccd.uncertainty, unit=ccd.unit)
            for ccd in self.ccds
        ]
        self.arrays = np.array([ccd.data for ccd in self.ccds])

    def profile(self):
        import line_profiler
        profile = line_profiler.LineProfiler(combine_arrays)
        profile.runcall(combine_arrays,
                        self.ccds,
                        method='mean',
                        clipping_method='sigclip')
        profile.print_stats()

    def ndcombine(self, **kwargs):
        combine_arrays(self.ccds, **kwargs)

    def ccdproc(self, **kwargs):
        ccdproc.combine(self.ccds, **kwargs)

    def dragons(self, **kwargs):
        stackit = NDStacker(**kwargs)
        stackit(self.ndds)

    def imcombinepy(self, **kwargs):
        imcombinepy.ndcombine(self.arrays, **kwargs)

    def measure_times(self, parallel=False):
        measures = defaultdict(dict)
        for code, bench in benchmarks.items():
            for name, params in bench.items():
                if name == 'ndcombine':
                    name = 'ndcombine parallel'
                    params['num_threads'] = 0 if parallel else 1
                func = time_execution(getattr(self, code))
                print(f'Running {code} - {name}', end=' ')
                measures[code][name] = func(**params)
                print()

        tbl = Table([{'package': name, **values}
                     for name, values in measures.items()])
        for col in tbl.itercols():
            if col.dtype.kind == 'f':
                col.format = '.2f'
        tbl.pprint(max_lines=-1, max_width=-1)

    def measure_memory(self):
        from memory_profiler import memory_usage
        measures = defaultdict(dict)
        for code, bench in benchmarks.items():
            for name, params in bench.items():
                print(f'Running {code} - {name}', end=' ')
                res = memory_usage(
                    (getattr(self, code), [], params),
                    timestamps=True,
                    interval=0.01,
                )
                measures[code][name] = np.max(np.array(res) - res[0])
                print()

        tbl = Table([{'package': name, **values}
                     for name, values in measures.items()])
        for col in tbl.itercols():
            if col.dtype.kind == 'f':
                col.format = '.2f'
        tbl.pprint(max_lines=-1, max_width=-1)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Run benchmarks')
    parser.add_argument('mode', nargs='?',
                        help='profile, memory, or cputime (default)')
    parser.add_argument('--parallel', action='store_true',
                        help='Use OpenMP (for ndcombine)')
    parser.add_argument('--limit', type=int)
    parser.add_argument('--datadir', type=str, default='~/data/combiner',
                        help='Directory where test data is stored')
    args = parser.parse_args()

    comp = Compare(limit=args.limit, datadir=args.datadir)
    if args.mode == 'profile':
        comp.profile()
    elif args.mode == 'memory':
        comp.measure_memory()
    else:
        comp.measure_times(parallel=args.parallel)


ndcombine/__init__.py
import numpy as np
from astropy.nddata import NDData, VarianceUncertainty

from .ndcombine import ndcombine
from .sigma_clip import sigma_clip  # noqa

try:
    from .version import version as __version__
except ImportError:
    __version__ = ''


def combine_arrays(
    data,
    mask=None,
    variance=None,
    method='mean',
    weights=None,
    clipping_method='none',
    clip_limits=(3, 3),
    # clipping methods
    # clip_extrema=False, nlow=1, nhigh=1,
    # minmax_clip=False, minmax_clip_min=None, minmax_clip_max=None,
    # sigma_clip=False, sigma_clip_low_thresh=3, sigma_clip_high_thresh=3,
    num_threads=0,
):
    """
    Parameters:
    -----------
    data : list of ndarray or list of NDData
        Data.
    method : str, {'mean', 'median', 'sum'}
        Combination method.
    clipping_method : str, {'minmax', 'extrema', 'sigmaclip', 'none'}
        Clipping method.

    """
    if isinstance(data[0], NDData):
        ndds = data
        data = np.asarray([nd.data for nd in ndds], dtype=np.float32)

        if ndds[0].mask is not None:
            # For now suppose that all NDData objects have a mask if the
            # first object has one.
            mask = np.asarray([nd.mask for nd in ndds], dtype=np.uint16)
        else:
            mask = None

        if ndds[0].uncertainty is not None:
            if not isinstance(ndds[0].uncertainty, VarianceUncertainty):
                raise ValueError('TODO')
            # For now suppose that all NDData objects have a mask if the
            # first object has one.
            variance = np.asarray([nd.uncertainty.array for nd in ndds],
                                  dtype=np.float32)
        else:
            variance = None
    else:
        data = np.asarray(data, dtype=np.float32)
        if mask is not None:
            mask = np.asarray(mask, dtype=np.uint16)
        if variance is not None:
            variance = np.asarray(variance, dtype=np.float32)

    shape = data.shape
    data = data.reshape(data.shape[0], -1)

    if variance is not None:
        variance = variance.reshape(data.shape[0], -1)

    if mask is None:
        mask = np.zeros_like(data, dtype=np.uint16)
    else:
        mask = mask.reshape(mask.shape[0], -1)

    lsigma, hsigma = clip_limits
    max_iters = 100

    outdata, outvar, outmask = ndcombine(
        data,
        mask,
        combine_method=method,
        hsigma=hsigma,
        lsigma=lsigma,
        max_iters=max_iters,
        num_threads=num_threads,
        reject_method=clipping_method,
        variance=variance,
    )

    outdata = outdata.reshape(shape[1:])
    if outvar is not None:
        outvar = VarianceUncertainty(outvar.reshape(shape[1:]))

    out = NDData(outdata, uncertainty=outvar)
    out.meta['REJMASK'] = outmask.reshape(shape)
    return out


benchmark.py
import os
import sys
import time
import timeit
from pathlib import Path

import bottleneck as bn
import numpy as np
from astropy.nddata import CCDData

from ndcombine import combine_arrays, ndcombine
from ndcombine.tests.helpers import make_fake_data

data = None
mask = None
datamasked = None


def test_median():
    global data, mask, datamasked

    shape = (10, 1000, 1000)
    np.random.seed(42)
    data = np.random.normal(size=shape).astype(np.float32)
    mask = np.zeros(shape, dtype=np.uint16)
    data = data.reshape(shape[0], -1)
    mask = mask.reshape(shape[0], -1)
    datamasked = np.ma.array(data, mask=mask.astype(bool))

    outndcomb, _, _ = ndcombine(data,
                                mask,
                                combine_method='median',
                                reject_method='none')
    outnp = np.median(data, axis=0)
    outnpma = np.ma.median(datamasked, axis=0)
    outbn = bn.median(data, axis=0)

    np.testing.assert_array_equal(outndcomb, outnp)
    np.testing.assert_array_equal(outndcomb, outnpma)
    np.testing.assert_array_equal(outndcomb, outbn)

    nb = 10
    kwargs = dict(globals=globals(), number=nb, repeat=5)

    def run(label, command):
        res = timeit.repeat(command, **kwargs)
        res = np.array(res) / nb
        print(f'{label:20s}: {np.mean(res):.3f}s ± {np.std(res):.3f}s')

    run('np.median', 'np.median(data)')
    run('np.ma.median', 'np.ma.median(datamasked, axis=0)')
    run('bn.median', 'bn.median(data)')
    run(
        'ndcombine 1 thread', "ndcombine(data, mask, combine_method='median', "
        "reject_method='none', num_threads=1)")
    run(
        'ndcombine', "ndcombine(data, mask, combine_method='median', "
        "reject_method='none')")


def get_ccddata():
    datadir = Path(os.path.expanduser('~/data/combiner'))

    if not datadir.exists():
        print('Creating test data')
        # datadir.mkdir(parents=True)
        make_fake_data(20, datadir, nsources=500)

    flist = list(datadir.glob('image-*.fits'))
    return [CCDData.read(f) for f in flist]


def profile(ccds):
    import line_profiler
    profile = line_profiler.LineProfiler(combine_arrays)
    profile.runcall(combine_arrays,
                    ccds,
                    method='mean',
                    clipping_method='sigclip')
    profile.print_stats()


def combine(ccds, n=5):
    t0 = time.time()
    for _ in range(n):
        print('.', end='', flush=True)
        combine_arrays(ccds,
                       method='mean',
                       clipping_method='sigclip',
                       num_threads=1)
    print('\nMean of 5 with 1 thread : {:.2f} sec.'.format(
        (time.time() - t0) / n))


def combine_parallel(ccds, n=5):
    t0 = time.time()
    for _ in range(n):
        print('.', end='', flush=True)
        combine_arrays(ccds,
                       method='mean',
                       clipping_method='sigclip',
                       num_threads=0)
    print('\nMean of 5 with max threads : {:.2f} sec.'.format(
        (time.time() - t0) / n))


if __name__ == "__main__":
    case = sys.argv[1] if len(sys.argv) > 1 else 'default'

    if case == 'median':
        test_median()
    else:
        ccds = get_ccddata()
        if case == 'profile':
            profile(ccds)
        elif case == 'parallel':
            combine_parallel(ccds)
        else:
            combine(ccds)
